Code/src/main/java/nl/utwente/viskell/ui/components/LambdaContainer.java

```diff
@@ -142,16 +149,49 @@ public final void handleConnectionChanges(boolean finalPhase) {
     }
     
     /** @return The local expression this LambdaContainer represents. */
-    public Expression getLocalExpr() {
-        List<Binder> binders = this.args.stream().map(arg -> arg.binder).collect(Collectors.toList());
-        LetExpression body = new LetExpression(this.res.getLocalExpr());
-        this.res.extendExprGraph(body);
-        return new Lambda(binders, body);
+    public Pair<Expression,Set<Block>> getLocalExpr() {
+        Pair<Expression, Set<Block>> pair = res.getLocalExpr();
+        List<Binder> binders = args.stream().map(arg -> arg.binder).collect(Collectors.toList());
+        LetExpression body = new LetExpression(pair.a, false);
+        Set<Block> surroundingBlocks = pair.b;
+        res.extendExprGraph(body, this, surroundingBlocks);
+        
+        return new Pair<>(new Lambda(binders, body), surroundingBlocks);
     }
 
     /** Called when the VisualState changed. */
     public void invalidateVisualState() {
         // TODO update anchors when they get a type label       
     }
+
+    @Override
+    public void attachBlock(Block block) {
+        attachedBlocks.add(block);
+        handleConnectionChanges(false);
+        handleConnectionChanges(true);
+    }
+
+    @Override
+    public boolean detachBlock(Block block) {
+        boolean removed = attachedBlocks.remove(block);
+        block.detachFromContainer();
+        handleConnectionChanges(false);
+        handleConnectionChanges(true);
```

Comment:
I've seen this (first call with false then call with true) a few times - what does it do? What happens if you forget one of the two steps? Can't the method just do both?